ggml.c

```diff
@@ -452,6 +480,89 @@ int64_t ggml_cycles_per_ms(void) {
 #define ggml_perf_cycles_per_ms() 0
 #endif
 
+//
+// NUMA support
+//
+
+struct ggml_numa_node
+{
+    uint32_t *cpus; // hardware threads on this node
+    uint32_t n_cpus;
+};
+
+struct ggml_numa_nodes
+{
+    struct ggml_numa_node *nodes;
+    uint32_t n_nodes;
+    uint32_t total_cpus; // hardware threads on system
+};
+
+struct ggml_numa_nodes ggml_numa = {
+    .nodes = NULL,
+    .n_nodes = 0,
+    .total_cpus = 0,
+};
+
+void ggml_numa_init(void)
+{
+    if (ggml_numa.n_nodes > 0) { return; }
+#ifdef __linux__
+    struct stat st;
+    char path[256];
+    int rv;
+    // enumerate nodes
+    while (true) {
+        rv = snprintf(path, sizeof(path), "/sys/devices/system/node/node%u", ggml_numa.n_nodes);
+        GGML_ASSERT(rv > 0 && (unsigned)rv < sizeof(path));
+        if (stat(path, &st) != 0) { break; }
+        ++ggml_numa.n_nodes;
+    }
+    // enumerate CPUs
+    while (true) {
+        rv = snprintf(path, sizeof(path), "/sys/devices/system/cpu/cpu%u", ggml_numa.total_cpus);
+        GGML_ASSERT(rv > 0 && (unsigned)rv < sizeof(path));
+        if (stat(path, &st) != 0) { break; }
+        ++ggml_numa.total_cpus;
+    }
+    GGML_PRINT_DEBUG("found %u numa nodes, %u CPUs\n", ggml_numa.n_nodes, ggml_numa.total_cpus);
+    if (ggml_numa.n_nodes < 1 || ggml_numa.total_cpus < 1) {
+        ggml_numa.n_nodes = 0;
+        return;
+    }
+    ggml_numa.nodes = calloc(ggml_numa.n_nodes, sizeof(struct ggml_numa_node));
```

Comment:
Avoid allocs or figure out a way to release the memory at the end of the program.
Alternatively, use static size arrays with max allowed size as constants